src/hotspot/share/gc/shenandoah/shenandoahPacer.cpp

```diff
@@ -249,32 +246,30 @@ void ShenandoahPacer::pace_for_alloc(size_t words) {
   JavaThread* current = JavaThread::current();
   if (current->is_attaching_via_jni() ||
       !current->is_active_Java_thread()) {
+    claim_for_alloc<true>(words);
     return;
   }
 
-  double start = os::elapsedTime();
-
-  size_t max_ms = ShenandoahPacingMaxDelay;
-  size_t total_ms = 0;
+  double const max_delay = static_cast<double>(ShenandoahPacingMaxDelay) / 1000;
+  double total_delay = 0;
 
-  while (true) {
+  double start = os::elapsedTime();
+  while (!claimed) {
```

Comment:
I suggest we common some exit paths by writing the loop like this:

```
double start_time = os::elapsedTime();
while (!claimed && (os::elapsedTime() - start_time) < max_delay) {
  // We could instead assist GC, but this would suffice for now.
  wait(1);
  claimed = claim_for_alloc<false>(words);
}
if (!claimed) {
  // Spent local time budget to wait for enough GC progress.
  // Force allocating anyway, which may mean we outpace GC,
  // and start Degenerated GC cycle.
  claimed = claim_for_alloc<true>(words);
  assert(claimed, "Should always succeed");
}
ShenandoahThreadLocalData::add_paced_time(JavaThread::current(), os::elapsedTime() - start_time);
```

Reply:
Thanks, refactored the code along with the change to use os::elapsed_counter(), only need handle the nanos to seconds conversion when calling  ShenandoahThreadLocalData::add_paced_time at the last.